packages/agreement-process/src/services/agreementContractBuilder.ts

```diff
@@ -1,35 +1,348 @@
-import { FileManager, Logger } from "pagopa-interop-commons";
-import { Agreement, EService, SelfcareId, Tenant } from "pagopa-interop-models";
-import { ApiAgreementDocumentSeed } from "../model/types.js";
+/* eslint-disable max-params */
+import path from "path";
+import { fileURLToPath } from "url";
+import {
+  FileManager,
+  Logger,
+  PDFGenerator,
+  dateAtRomeZone,
+  formatDateyyyyMMddHHmmss,
+  timeAtRomeZone,
+} from "pagopa-interop-commons";
+import {
+  Agreement,
+  AgreementContractPDFPayload,
+  AgreementDocumentId,
+  Attribute,
+  CertifiedTenantAttribute,
+  DeclaredTenantAttribute,
+  EService,
+  SelfcareId,
+  Tenant,
+  TenantAttributeType,
+  TenantId,
+  VerifiedTenantAttribute,
+  UserId,
+  generateId,
+  genericError,
+  tenantAttributeType,
+} from "pagopa-interop-models";
+import {
+  UserResponse,
+  selfcareV2Client,
+} from "pagopa-interop-selfcare-v2-client";
+import { match } from "ts-pattern";
+import {
+  agreementMissingUserInfo,
+  agreementStampNotFound,
+  userNotFound,
+} from "../model/domain/errors.js";
 import { UpdateAgreementSeed } from "../model/domain/models.js";
-import { pdfGenerator } from "./pdfGenerator.js";
+import { ApiAgreementDocumentSeed } from "../model/types.js";
+import { AgreementProcessConfig } from "../utilities/config.js";
 import { AttributeQuery } from "./readmodel/attributeQuery.js";
 
+const retrieveUser = async (
+  selfcareId: SelfcareId,
+  id: UserId
+): Promise<UserResponse> => {
+  const user = await selfcareV2Client.getUserInfoUsingGET({
+    queries: { institutionId: selfcareId },
+    params: { id },
+  });
+
+  if (!user) {
+    throw userNotFound(selfcareId, id);
+  }
+  return user;
+};
+
+const createAgreementDocumentName = (
+  consumerId: TenantId,
+  producerId: TenantId
+): string =>
+  `${consumerId}_${producerId}_${formatDateyyyyMMddHHmmss(
+    new Date()
+  )}_agreement_contract.pdf`;
+
+const getAttributeInvolved = async (
+  consumer: Tenant,
+  seed: UpdateAgreementSeed,
+  attributeQuery: AttributeQuery
+): Promise<{
+  certified: Array<[Attribute, CertifiedTenantAttribute]>;
+  declared: Array<[Attribute, DeclaredTenantAttribute]>;
+  verified: Array<[Attribute, VerifiedTenantAttribute]>;
+}> => {
+  const getAgreementAttributeByType = async <
+    T extends
+      | CertifiedTenantAttribute
+      | DeclaredTenantAttribute
+      | VerifiedTenantAttribute
+  >(
+    type: TenantAttributeType
+  ): Promise<Array<[Attribute, T]>> => {
+    const seedAttributes = match(type)
+      .with(tenantAttributeType.CERTIFIED, () => seed.certifiedAttributes || [])
+      .with(tenantAttributeType.DECLARED, () => seed.declaredAttributes || [])
+      .with(tenantAttributeType.VERIFIED, () => seed.verifiedAttributes || [])
+      .exhaustive()
+      .map((ca) => ca.id);
```

Comment:
```suggestion
      .map((attribute) => attribute.id);
```

Reply:
Added with commit [291ef33](https://github.com/pagopa/interop-be-monorepo/pull/520/commits/291ef33f184779c13fdb684193b096ee519e49c4)